projects/telemetry/src/gps.c

```diff
@@ -0,0 +1,78 @@
+#include <stdio.h>
+#include <string.h>
+#include <stdint.h>
+#include "status.h"
+#include "gps.h"
+#include "gpio.h"
+#include "uart_mcu.h"
+#include "misc.h"
+#include "nmea.h"
+
+// How many handlers can we possibly need?
+#define GPS_HANDLER_ARRAY_LENGTH 5
+
+static const UARTPort port = UART_PORT_2;
+static GPSHandler gps_handler[GPS_HANDLER_ARRAY_LENGTH] = {0};
+static GGAHandler gga_handler[GPS_HANDLER_ARRAY_LENGTH] = {0};
+static uint32_t init = 0;
+
+int32_t add_gps_handler(GPSHandler handler) {
+  for (uint32_t i = 0; i < GPS_HANDLER_ARRAY_LENGTH; i++) {
+    if (gps_handler[i] == NULL) {
+      gps_handler[i] = handler;
+      // This cast won't be an issue, there's no way "i" will be big enought to cause an overflow
+      return (int32_t) i;
+    }
+  }
+  return -1;
+}
+
+int32_t add_gga_handler(GGAHandler handler) {
+  for (uint32_t i = 0; i < GPS_HANDLER_ARRAY_LENGTH; i++) {
+    if (gga_handler[i] == NULL) {
+      gga_handler[i] = handler;
+      // This cast won't be an issue, there's no way "i" will be big enought to cause an overflow
+      return (int32_t) i;
+    }
+  }
+  return -1;
+}
+
+void remove_gps_handler(uint32_t index) {
+  if (index >= GPS_HANDLER_ARRAY_LENGTH) return;
```

Comment:
Usually we make our guard clauses return a status. It helps with debugging. Same throughout the rest of this file.